orttraining/orttraining/training_ops/cuda/communication/recv.h

```diff
@@ -4,30 +4,25 @@
 #ifdef USE_HOROVOD
 
 #pragma once
-#include <mpi.h>
 #include "core/common/common.h"
 #include "core/providers/cuda/cuda_common.h"
 #include "core/providers/cuda/cudnn_common.h"
-#include "common.h"
 
 namespace onnxruntime {
 namespace cuda {
 
-template <typename T>
 class Recv final : public CudaKernel {
 public:
   Recv(const OpKernelInfo& info) : CudaKernel(info) {
     ORT_ENFORCE(info.GetAttr<int64_t>("tag", &tag_).IsOK());
-    ORT_ENFORCE(info.GetAttr<int64_t>("src", &src_).IsOK());
-    ORT_ENFORCE(info.GetAttr<int64_t>("dst", &dst_).IsOK());
+    ORT_ENFORCE(info.GetAttrs<int64_t>("element_types", element_types_).IsOK());
```

Comment:
it seems the use of element_types_ is to get the total number of tensors. Consider change this into a int type instead of a vector. 

Same for rend.h

Reply:
It is used for type_inference in schema.